### PR TITLE
Share permalink feature

### PR DESCRIPTION
**Pull request by olivierSemet**
Related to https://github.com/camptocamp/c2cgeoportal/issues/2071

**Demo:** 
- Desktop: https://oliviersemet.github.io/ngeo/share-permalink-feature/examples/contribs/gmf/apps/desktop/

- Desktop alt: https://oliviersemet.github.io/ngeo/share-permalink-feature/examples/contribs/gmf/apps/desktop_alt/

**Important**

- Currently the call to the Shortener API does not work (HTTP Bad request) because of a security reason. Indeed, the server check if the URL sended to be shortened is part of the domain  `geomapfish-demo.camptocamp.net` . Therefore it cannot work on my gh pages. @sbrunner will work on it.

- The _Desktop_ app use the Shortener API whereas the _Desktop Alt_ app does not (Testing the case when no Shortener API is available for the customer).

- In case of error, I display the following message ` Error, cannot get the shortened URL.` I'm not sure this is a good message for the customer. Please fill free to suggest an alternative.

- Despite the error when trying to get a short URL, a first review of the code can be done now.

**Test**

- [ ]  Shortener API calls are triggered only when the modal is shown. (Desktop)

- [ ]  Cannot share the shortened url as long as the email provided is invalid (Desktop)

- [ ]  In case of error when trying to get a short URL, the permalink field is filled with the original URL + error message (Desktop)

- [ ] A valid URL must be shortened (Desktop) _(currently not available)_

- [ ]  the permalink field is filled with the original URL because there is no Shortener API available  (Desktop Alt)

- [ ] A warning must be displayed when a very long URL (length > 2083) is provided (Both apps) 

@pgiraud @fredj @sbrunner could you have a look please?

### CODE REVIEW

File: contribs/gmf/apps/mobile/index.html [hunk 1]
```diff
@@ -179,6 +179,7 @@
         module.constant('fulltextsearchUrl', 'https://geomapfish-demo.camptocamp.net/2.1/wsgi/fulltextsearch?limit=20');
         module.constant('gmfWmsUrl', 'https://geomapfish-demo.camptocamp.net/2.1/wsgi/mapserv_proxy');
         module.constant('gmfTreeUrl', 'https://geomapfish-demo.camptocamp.net/2.1/wsgi/themes?version=2&background=background');
+        module.constant('shortenerCreateURL', '');
```
Comment (olivierSemet) [inline contribs/gmf/apps/mobile/index.html]: should be instead of gmfShortenerCreateUrl

File: contribs/gmf/externs/gmfx.js [hunk 2]
```diff
@@ -511,3 +514,36 @@ gmfx.TimePropertyResolutionEnum = {
  * }}
  */
 gmfx.TimeProperty;
+
+
+/**
+ * @typedef {{
+ *  data: Object,
```
Comment (fredj) [inline contribs/gmf/externs/gmfx.js]: Can you use `{{ short_url: string }}` instead of `Object`?

Comment (olivierSemet) [inline contribs/gmf/externs/gmfx.js]: I define later `gmfx.ShortenerAPIResponse.data`  as `@type {{ short_url : string }}` line 532.
I've never seen a record used inside a typedef but I'll try it

Comment (olivierSemet) [inline contribs/gmf/externs/gmfx.js]: Not possible: `contribs/gmf/externs/gmfx.js:521: ERROR - Bad type annotation. type not recognized due to syntax error
ERR! compile  *  data: {{ short_url : string }},`

The best I can do is to replace `Object` by `Object.<string>`

Comment (fredj) [inline contribs/gmf/externs/gmfx.js]: Or create at typedef for gmfx.ShortenerAPIResponse.data

File: contribs/gmf/externs/gmfx.js [hunk 3]
```diff
@@ -511,3 +514,36 @@ gmfx.TimePropertyResolutionEnum = {
...
+ * }}
+ */
+gmfx.ShortenerAPIResponse;
+
+
+/**
+ * Response payload to the shortener API
+ * @type {{ short_url : string }}
+ */
+gmfx.ShortenerAPIResponse.data;
```
Comment (olivierSemet) [inline contribs/gmf/externs/gmfx.js]: Here

File: contribs/gmf/externs/gmfx.js (hunk 2 above)
Comment (olivierSemet) [inline contribs/gmf/externs/gmfx.js]: Yes it was already done

File: contribs/gmf/externs/gmfx.js (hunk 3 above)
Comment (fredj) [inline contribs/gmf/externs/gmfx.js]: see https://github.com/olivierSemet/ngeo/pull/2

**Comment by fredj** [comment]
What about merging this pull request and create an issue with the test section (checkboxes) in the description.
When the shortener server is fixed, we can validate and fix if needed.

**Comment by fredj** [comment]
> What about merging this pull request and create an issue with the test section (checkboxes) in the description.

see #1554

**Comment by fredj** [comment]
@olivierSemet please rebase to master

**Comment by olivierSemet** [comment]
@fredj should be good after Jenkins checks.

**Pull request merged**